### PR TITLE
Add WAGTAIL_APPEND_SLASH setting

### PR DESCRIPTION
**Pull request by atbaker**
Hi guys,

This PR addresses #2096. It's incomplete — it needs tests and docs changes. But since adding a new setting isn't a trivial change, I wanted to start this PR so we could agree on the best way to do this.

We're getting ready to launch a new Wagtail-backed service at work, serving only a few pages of content at first but gradually transitioning off our old platform. The old platform doesn't append slashes, so to keep consistency across the site during the transition we'll be stripping them from our new Wagtail URLs too. Hence, our need for a solution to #2096.

**A summary of my first stab:**

- Set a new constant in wagtailcore.utils, `WAGTAIL_APPEND_SLASH`, which defaults to `True` if it's not present in the Django settings
- Reference that constant in wagtailcore.urls, using different regex's for the `wagtail_serve` URL
    - The new regex simply makes trailing slashes optional. That seemed like the best behavior to mimic Django's `APPEND_SLASH` setting, which I would guess is what most other devs would expect
- Django's `reverse` always returns a trailing slash, so to make links appear without trailing slahes I had to add some extra logic to the Page model's `get_url_parts` method

Happy to tweak this whichever way you think is best. Once we agree on an approach, I'll fill in the tests and add some notes to the docs.

Wagtail has been a delight to work with these past couple months. Thank you for creating, sharing, and maintaining such a fantastic library.

### CODE REVIEW

**Comment by m1kola** [comment]
@atbaker, thank you for your participation!

At first glance it looks good, but I have doubts about how `WAGTAIL_APPEND_SLASH` should work with Django's `APPEND_SLASH` (I mean in various combinations).

File: wagtail/wagtailcore/urls.py [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
 from wagtail.wagtailcore import views
+from wagtail.wagtailcore.utils import WAGTAIL_APPEND_SLASH
+
+
+# If WAGTAIL_APPEND_SLASH is False, allow Wagtail to serve pages on URLs
+# with and without trailing slahes
+if WAGTAIL_APPEND_SLASH:
+    serve_pattern = r'^((?:[\w\-]+/)*)$'
+else:
+    serve_pattern = r'^((?:[\w\-]+/?)*)$'
```
Comment (m1kola) [inline wagtail/wagtailcore/urls.py]: Looks like this pattern will be very slow.

For example you can try this:
```python
import re
re.match(r'^((?:[\w\-]+/?)*)$', 'media/images/James_Joyce_in_1915.width-300_6M5Jvbh.jpg')
```

**Comment by atbaker** [comment]
@m1kola - Thanks for the first review!

You were right about that regex - it had a catastrophic backtrack. I updated it to be simpler and more efficient.

Regarding the `WAGTAIL_APPEND_SLASH` setting's interaction with Django's `APPEND_SLASH`, here's how it works:

<br/> | WAGTAIL_APPEND_SLASH = True (default) | WAGTAIL_APPEND_SLASH = False
------------ | ------------- | -----------
**APPEND_SLASH = True (default)** | Original Wagtail behavior: requests to URLs without trailing slashes are caught by `CommonMiddleware`, which 301 redirects to a version with a trailing slash which Wagtail serves | Wagtail pages can be reached with and without trailing slashes. No redirects occur.
**APPEND_SLASH = False** | Requests that include a trailing slash are served. Requests without it are not redirected by `CommonMiddleware` and are 404'd | Wagtail pages can be reached with and without trailing slashes. No redirects occur.

If you think this is looking good, I'll update the docs and do one last round of polish on the PR.

**Comment by gasman** [comment]
Thanks! Merged in 228088d33d3d36afdc8363f9e59f1a8cbcf6d072.

**Comment by atbaker** [comment]
Thanks for shepherding this along! Hoping this is the first of many PRs to the project.

**Pull request closed without merging**